### PR TITLE
Support temporary AWS credential (STS)

### PR DESCRIPTION
**Pull request by nicholasren**
Some times I need to supply a temporary AWS credentail to logstash plugins(e.g. SNS) for testing purpose. so I created this patch. Also I refactor the specs.

### CODE REVIEW

File: lib/logstash/plugin_mixins/aws_config.rb [hunk 1]
```diff
@@ -56,25 +59,21 @@ def aws_options_hash
...
+    if @access_key_id && @secret_access_key
+      opts = {
+        :access_key_id => @access_key_id,
+        :secret_access_key => @secret_access_key
+      }
+      opts[:session_token] = @session_token if @session_token
+    elsif @aws_credentials_file
+      opts = YAML.load_file(@aws_credentials_file)
+    else
+      fail 'Neither credential file and credential was provided'
```
Comment (ph) [inline lib/logstash/plugin_mixins/aws_config.rb]: I would raise a `LogStash::ConfigurationError` error instead of using `fail`.

File: spec/plugin_mixin/aws_config_spec.rb [hunk 2]
```diff
@@ -14,25 +14,52 @@ def aws_service_endpoint(region)
...
-  it 'should support reading configuration from a yaml file' do
-    settings = { 'aws_credentials_file' => File.join(File.dirname(__FILE__), '..', 'fixtures/aws_credentials_file_sample_test.yml') }
-    config = DummyInputAwsConfig.new(settings)
-    config.aws_options_hash[:access_key_id].should == '1234'
-    config.aws_options_hash[:secret_access_key].should == 'secret'
+  context 'inline credential' do
+    context 'temporary credential' do
+      let(:settings) { { 'access_key_id' => '1234', 'secret_access_key' => 'secret', 'session_token' => 'session_token' } }
+
+      it "should support passing as key, value, and sesson_token" do
```
Comment (ph) [inline spec/plugin_mixin/aws_config_spec.rb]: small typo here: `sesson_token` should be `session_token`

File: spec/plugin_mixin/aws_config_spec.rb [hunk 3]
```diff
@@ -14,25 +14,52 @@ def aws_service_endpoint(region)
...
+          subject[:access_key_id].should == settings['access_key_id']
+          subject[:secret_access_key].should == settings['secret_access_key']
+        end
+      end
+    end
   end
 
-  it 'should call the class to generate the endpoint configuration' do
-    settings = { 'access_key_id' => '1234',  'secret_access_key' => 'secret', 'region' => 'us-west-2' }
+  context 'regoin provided' do
```
Comment (ph) [inline spec/plugin_mixin/aws_config_spec.rb]: small typo: `regoin` => `region`

File: lib/logstash/plugin_mixins/aws_config.rb (hunk 1 above)
Comment (jordansissel) [inline lib/logstash/plugin_mixins/aws_config.rb]: STORY TIME.

I'm pretty sure `fail` and `raise` in ruby do the same thing. But yes, we should raise an error (we never use fail, mostly for style consistency reasons), and a LogStash::COnfigurationError is a good one to use!

`raise LogStash::ConfigurationError, https://github.com/elasticsearch/logstash-forwarder/pull/327`

Comment (ph) [inline lib/logstash/plugin_mixins/aws_config.rb]: Yes they do the same thing.  The link is a cut and paste hiccup?

**Comment by nicholasren** [comment]
yes, it make perfect sense to raise a `LogStash::ConfigurationError` here, code changed.

**Comment by ph** [comment]
Thank you @nicholasren, I'll do more test on my side and get back to you.
Thanks for the spec refactoring, a few good ideas there.

**Comment by nicholasren** [comment]
Hi @ph any update on this?

**Comment by ph** [comment]
Correct me if I am wrong, this is my understanding of the AWS STS.

1. You run some piece of code in a trusted place with your `ACCESS_ID` and your `SECRET_KEY` to generate a STS Token valid for a predefined time. (iirc using the web ui give you a 1 hour duration)

2. Using that token you don't need to have the ACCESS_ID and `SECRET_KEY` in your configuration? So we could remove the validation and fallback to the STS token.

3. If you export`AWS_SECURITY_TOKEN` in the environment the aws-sdk should pick it up and use it.

**Comment by nicholasren** [comment]
FYI: http://docs.aws.amazon.com/STS/latest/UsingSTS/using-temp-creds.html
> 1. You run some piece of code in a trusted place with your ACCESS_ID and your SECRET_KEY to generate a STS Token valid for a predefined time. (iirc using the web ui give you a 1 hour duration)

     actually I'll get a temporary credential which including `ACCESS_ID`, `SECRET_KEY` and a `SECURITY_TOKEN` .

> 2.  Using that token you don't need to have the ACCESS_ID and SECRET_KEY in your configuration? So we could remove the validation and fallback to the STS token.

    No, you till need the `ACCESS_ID`, `SECRET_KEY` and a `SECURITY_TOKEN` from generated temporary credential to be able to access AWS resources

> 3. If you exportAWS_SECURITY_TOKEN in the environment the aws-sdk should pick it up and use it.

    Yes

**Comment by ph** [comment]
Thanks for the clarification and the good work!

I have tested it with my S3 input/output config and checked the aws-sdk code.
LGTM!

Would you mind doing step 2 of the contributing guide? https://github.com/elasticsearch/logstash/blob/master/CONTRIBUTING.md#contribution-steps

**Comment by nicholasren** [comment]
Yes, happy to do that.
CLA signed. :)

**Comment by ph** [comment]
@nicholasren I've checked our CLA database and I found a @nicholasren but not the with the same email than your patch see https://github.com/logstash-plugins/logstash-mixin-aws/pull/4.patch

Could you do either:
1. Sign the CLA with the email in your patch. (probably the easiest one)
2. or change the commit email and pushed it again?

I agree it is a little cumbersome but we are using this information to track all the commits with our automated tools.

File: lib/logstash/plugin_mixins/aws_config.rb [hunk 4]
```diff
@@ -56,25 +59,21 @@ def aws_options_hash
...
+    if @access_key_id && @secret_access_key
+      opts = {
+        :access_key_id => @access_key_id,
+        :secret_access_key => @secret_access_key
+      }
+      opts[:session_token] = @session_token if @session_token
+    elsif @aws_credentials_file
+      opts = YAML.load_file(@aws_credentials_file)
+    else
+      raise LogStash::ConfigurationError.new('Neither credential file and credential was provided')
```
Comment (ph) [inline lib/logstash/plugin_mixins/aws_config.rb]: @nicholasren I was reviewing old tickets concerning the AWS and I did a double check on this PR.
the raise `LogStash::ConfigurationError` will break defining the AWS configuration in the shell environment. 

We could add a default to the config to load the configuration from the environment.

```ruby
config : access_key_id, :default => ENV['AWS_ACCESS_KEY_ID']
config : access_key_id, :default => ENV['AWS_SECRET_ACCESS_KEY']
config : access_key_, :default => ENV['AWS_SESSION_TOKEN']
```

Sadly we don't have a test to cover this case :(

Comment (ph) [inline lib/logstash/plugin_mixins/aws_config.rb]: It was working as a side effect of using aws-sdk-ruby`, the gem fallback to the environment if no credentials are provided.

Comment (ph) [inline lib/logstash/plugin_mixins/aws_config.rb]: If the ruby sdk cannot find the configuration in the environment it will fallback to IAM Role.

```
specified credentials -> environment -> ec metadata.
```

Comment (nicholasren) [inline lib/logstash/plugin_mixins/aws_config.rb]: Hi @ph,

Yes, got it.
The responsibility of this plugin is to enable programmer to specify aws credentials before using any logstash aws related plugin, (ethier inline or a credential file). 

we __shouldn't__ stop `ruby-aws-sdk` from using the credential in environment variable and ec2 instance metadata. 

I will remove the related code.

Comment (nicholasren) [inline lib/logstash/plugin_mixins/aws_config.rb]: I didn't add a test like `should use system variable while no credential provided` here, because I think it's not this plugin's responsibility.  it is how the `ruby-aws-sdk` works. if really don't want this plugin be break later, we could add this test as an integration test in other plugins like `logstash-output-sns` or `logstash-output-s3`.

Comment (ph) [inline lib/logstash/plugin_mixins/aws_config.rb]: :+1: 
I think the best way would be to create a shared example in this repo and use it in the other plugins. Your thoughts?

This could be another PR.

**Comment by ph** [comment]
I am comfortable with the changes and the CLA is signed, could you rebase it?

**Comment by nicholasren** [comment]
rebased :smile:
